[Hamming]: Approaches (#3304)
exercism/python: exercises/practice/hamming/.approaches/sum/content.md

# sum

The benefit of using [`sum()`][sum] is that we can use a list comprehension to create a list of booleans.
Then we can pass that list to `sum()` and it will add up all the booleans.
Where `True` is treated as 1 and `False` is treated as 0.
Then that sum is returned.

This can make the code a bit more concise.

Here is an example using `sum()` and `zip()`:

```python
def distance(strand_a, strand_b):
    if len(strand_a) != len(strand_b):
        raise ValueError("Strands must be of equal length.")
    return sum(nucleotide_a != nucleotide_b for nucleotide_a, nucleotide_b in zip(strand_a, strand_b))
```

This approach starts by checking if the two strands are of equal length by using [`len()`][len].
If not, a [`ValueError`][value-error] is raised.

After that is checked, a variable `count` is initialized to 0.
The count variable will be used to keep track of the number of differences between the two strands.

This approach uses the [`zip()`][zip] function to iterate over two strings.

> You can read more about how to solve this exercise with `zip()` in the [zip approach][approach-zip].

Reviewer: ```suggestion
You can read more about how to solve this exercise with `zip` in the [zip approach][approach-zip].
```

> What differs in this approach is that we use a list comprehension to create a list of booleans.

Reviewer: ```suggestion
What differs in this approach is that we use a generator expression to create booleans.
```

> The list comprehension iterates over the tuples returned by `zip()`.

Reviewer: ```suggestion
The generator expression returns an iterator over the tuples returned by `zip`.
```

> Then under the iteration so are the tuples unpacked into two variables, `nucleotide_a` and `nucleotide_b`.

Reviewer: ```suggestion
Within the iteration, the tuples are unpacked into two variables, `nucleotide_a` and `nucleotide_b`.
```

> We then compare the characters `nucleotide_a` and `nucleotide_b`.

Reviewer: ```suggestion
We can then compare `nucleotide_a` and `nucleotide_b`.
```

> If they are not equal, we add `True` to the list.

Reviewer: ```suggestion
If they are **not** equal, `True` is produced.
```

> If they are equal, we add `False` to the list.

Reviewer: ```suggestion
If they **are** equal, `False` is produced.
```

> The list comprehension is then passed to the `sum()` function.

Reviewer: ```suggestion
The generator expression is then passed to the `sum` function.
```

> The [`sum()`][sum] function will add up all the booleans in the list.

Reviewer: ```suggestion
`sum` will then iterate over the generator expression and add up all the booleans.
```

Where `True` is treated as 1 and `False` is treated as 0.
You can read more about this behavior in [Boolean as numbers][booleans].

> Then that sum is returned.

Reviewer: ```suggestion
Finally the totaled booleans are returned.
```

> This approach is also doable with range but it is a bit more verbose:

Reviewer: ```suggestion
This approach is also doable with `range` but it is a bit more verbose:
```
